### ISSUE TITLE
Version 3 breaks vite proxy

### ISSUE THREAD

**Issue by xrado** [vbenjs/vite-plugin-html#38]
In version v2.7.13 all worked fine, after upgrading to v3 (3.0.6) vite server proxy settings do work anymore if accessed directly via browser. Lets say I have in my vite config: 
```
server: {
	proxy: {
		'/api': 'http://localhost:3000'
	},
}
```
calling localhost:8000/api/list directly from my browser always serves index.html only async XHR calls after initial load goes through the proxy.

**Comment by anncwb** [comment]
can you give me a reproduction？

**Comment by xrado** [comment]
@anncwb here https://stackblitz.com/edit/vitejs-vite-j1uvi4?file=vite.config.js&terminal=dev

open and add `/api` to the end of preview url, enter and click the preview reload button
empty page, does not load because tries to load `http://localhost:3000` via proxy

uncomment createHtmlPlugin, save and try again
`/api` always loads `index.html` because vite proxy is ignored or overriden somehow

**Comment by oscartbeaumont** [comment]
I am having this same issue but it was a bit confusing to work out because the proxying only fails when the client accepts `text/html`, so some of my API requests were getting through.

This is being caused by the [configureServer](https://vitejs.dev/guide/api-plugin.html#configureserver) Vite plugin API ([used here](https://github.com/vbenjs/vite-plugin-html/blob/b1427bd59df1f2879c31fc073bedfbb60c5f479d/packages/core/src/htmlPlugin.ts#L52)) which is catching all HTTP requests if the client accepts `text/html` hence preventing Vite from proxying them.

I am not sure why you are using the `configureServer` API in the first place because `transformIndexHtml` is run by Vite when the devserver is enabled which should achieve the functionality you need. I think it's worth looking at removing it or at least making it an optional (disabled by default) feature as this custom devserver will probably introduce more issues down the road as Vite make internal changes even if you were to make a workaround for this specific issue.

**Comment by oscartbeaumont** [comment]
I am also happy to PR the change if requested.

**Comment by anncwb** [comment]
PR welcome

**Comment by oscartbeaumont** [comment]
It looks like this change was fixed in release 3.2.0. @anncwb or @xrado can we get this issue closed.

**Comment by xrado** [comment]
still doesn't solve the issue for me. Vite proxy settings are still overridden by the plugin on the initial load of the page.
Lets say I want to open api endpoint url in the new tab to see the response, instead of the response index.html loads.

**Comment by Renderz** [comment]
> It looks like this change was fixed in release 3.2.0. @anncwb or @xrado can we get this issue closed.

still not working.

**Comment by ex0ns** [comment]
I also have issue with the vite proxy whenever `vite-plugin-html` is used in the configuration (see https://github.com/vitejs/vite/issues/7999).
As explained on this issue this is only happening when the header `Accept: html` is present

https://stackblitz.com/edit/vitejs-vite-2chcqw?file=index.html 
```
yarn vite serve --debug
```

In `index.html` you can try to comment/uncomment the `Accept` header line
In `vite.config.js` if you remove/comment the html plugin, this works fine as well

I guess this issue might be out of the scope of vite ? I don't know enough about plugins  to understand how they could have such an influence on the server and proxy

**Comment by erm1116** [comment]
I also have same issue. 
And trying to remove below lines, it works fine.
https://github.com/vbenjs/vite-plugin-html/blob/841d4ef04c3cf5ff0d4339350ae336aa83aa70ed/packages/core/src/htmlPlugin.ts#L87-L93

I don't clearly understand why these lines are needed. (maybe, [history mode support](https://github.com/vbenjs/vite-plugin-html/commit/20a7c69ed7f8f355bda923dd9f84717727276c67) for vue-router ?)
Can't we remove this?

**Comment by vellengs** [comment]
同样遇到这个问题，导致接口的查询参数丢失。

**Comment by JackieCheung** [comment]
same problem, any progress now?

**Comment by kilobyte2007** [comment]
Hey, we have the same problem. It seems the plugin drops all of the query parameters that are passed when accessed via a browser (so with the accepts headers). I am not sure entirely why the middleware with the htmlAcceptHeaders is needed so it's hard to think of a fix but if someone can explain it I'd be happy to make a PR.

**Comment by MetaMmodern** [comment]
@anncwb this is not related to only proxy requests, this issue covers all requests that are made to an application being run through `vite serve`. Due to this I can't open any asset that's being loaded in a separate tab, the link leading to a file simply opens index.html instead and the website loads.
Idk what was the reason for that, but I'd be glad to help with a PR, since we use this plugin to inject app urls for og metatags and this bug blocks and old implemented feature of our website.

**Comment by moroz1k** [comment]
same problem

**Comment by ehossack** [comment]
Also have this problem, and tested locally that this fix does indeed work:

https://github.com/vbenjs/vite-plugin-html/pull/96

@anncwb mind taking a look at the PR?

**Comment by marcovdb** [comment]
Also experienced this issue today! Hope the PR can be merged soon.

**Comment by snecker** [comment]
Any update?

**Comment by erg** [comment]
This pr (https://github.com/vbenjs/vite-plugin-html/pull/96) solved it for me.

I am proxying port 8080 to a react backend on port 3000 with `vite-plugin-html`.

If I try my url on port 8080 from Chrome then the GET params are dropped by the vite proxy; they do not make it to the backend. If I use `curl` then it works as expected. As someone else mentioned, it is probably that the Chrome headers are triggering a behavior header with `Accept: text/html` which Chrome sends by default and `curl` does not send.

Chrome `114.0.5735.133` sends `Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7` by default.

If I export the API call from Chrome as a `fetch` and call it with the `Accept: application/json` then the vite proxy works as expected as well.

```
chrome browser url address:
  http://localhost:8080/proxy/lists/getListById.sql?listId=30747
backend sees:
  /proxy/epi/epiquery1/glglive/lists/v2/getListById.sql # WRONG

chrome browser url address:
  http://localhost:3000/proxy/lists/getListById.sql?listId=30747
backend sees:
  /proxy/epi/epiquery1/glglive/lists/v2/getListById.sql?listId=30747

curl:
  curl http://localhost:8080/proxy/lists/getListById.sql?listId=30747
backend sees:
  http://localhost:8080/proxy/lists/getListById.sql # WRONG

curl:
  curl http://localhost:3000/proxy/lists/getListById.sql?listId=30747
backend sees:
  http://localhost:3000/proxy/lists/getListById.sql?listId=30747

# FAILS
fetch('http://localhost:8080/proxy/getListById.sql?listId=30747', {headers: {'Accept': 'text/html'}})
  .then(response => response.json())
  .then(data => console.log(data));

# WORKS
fetch('http://localhost:8080/proxy/getListById.sql?listId=30747', {headers: {'Accept': 'application/json'}})
  .then(response => response.json())
  .then(data => console.log(data));

# WORKS (no Accept header)
fetch('http://localhost:8080/proxy/getListById.sql?listId=30747&abc=123')
  .then(response => response.json())
  .then(data => console.log(data));

# All three of the fetch calls work on port 3000
```

Applying the patch from https://github.com/vbenjs/vite-plugin-html/pull/96 solves this--port 8080 starts forwarding the GET params.

I'm not using the `configureServer` plugin.

The vite setup in question does not use anything weird:

```javsacript
import react from '@vitejs/plugin-react';
import 'dotenv/config.js';
import path from 'path';
import { defineConfig, splitVendorChunkPlugin } from 'vite';
import { createHtmlPlugin } from 'vite-plugin-html';

const REACT_APP_PATH_ROOT = process.env.REACT_APP_PATH_ROOT || '{{{REACT_APP_PATH_ROOT}}}';

export default defineConfig({
  server: {
    port: 8080,
    base: REACT_APP_PATH_ROOT,
    proxy: {
      '/proxy': `http://localhost:3000`
    }
  },
  plugins: [
    splitVendorChunkPlugin(),
    createHtmlPlugin({
      pages: [
        {
          inject: {
            data: {
              reactAppPathRootClean: REACT_APP_PATH_ROOT,
              reactAppPathRoot: REACT_APP_PATH_ROOT
            }
          }
        },
        {
          filename: path.resolve(__dirname, 'src/loginGoodbye.html')
        }
      ]
    }),
    react()
  ],

  ... // more config
}
```

Using `"vite-plugin-html": "^3.2.0"` and  `"vite": "^4.3.9"`.

If you want a minimal test project I could throw one together I suppose.

TLDR: please merge https://github.com/vbenjs/vite-plugin-html/pull/96

**Comment by LvChengbin** [comment]
How does such a terrible bug exist for years?

**Comment by kilobyte2007** [comment]
@LvChengbin yep, frustrating. There are PRs to fix this but it looks like the repo is unmaintained. I'll probably have to fork it and release it on npm to use in our own project.

**Comment by PeteTheHeat** [comment]
I found this issue because I was seeing similar symptoms: `vite-plugin-html` was breaking my vite server proxy, specifically when I was using a Regex https://vitejs.dev/config/server-options.html#server-proxy

The solution was to patch the `createRewire` function here https://github.com/vbenjs/vite-plugin-html/blob/main/packages/core/src/htmlPlugin.ts#L357

```
 const isApiUrl = proxyUrlKeys.some((item) => {
        if (item.startsWith("^")){
          return new RegExp(item).test(pathname)
        }
        return pathname.startsWith(path.resolve(baseUrl, item))
      });
```

**Comment by Sjaaky** [comment]
I've just added this hack to our vite.config.ts counter this problem.
```
const fixIndexHtmlRedirect = (plugins: PluginOption[]): PluginOption[] => {
  return plugins.map((plugin) => {
    if (!plugin) return plugin;
    if (Array.isArray(plugin)) {
      return fixIndexHtmlRedirect(plugin as PluginOption[]);
    } else if (typeof plugin === 'object' && 'name' in plugin && plugin.name === 'vite:html') {
      delete (plugin as Plugin).configureServer; // this line is where the magic happens
    }
    return plugin;
  });
};

export default defineConfig({
  base: '',
  plugins: fixIndexHtmlRedirect([
    vue(),
    ...htmlPlugins,
    htmlPluginGtm(gtmId),
   ..etc..

```
Disclaimer: Typescript/javascript is not my native tongue. The detection of vite:html plugins could probably be improved.
